### ISSUE TITLE
Exponentiation with Rat or FatRat exponent results in Num

### ISSUE THREAD

**Issue by landyacht** [Raku/problem-solving#209]
If the exponent passed to `infix:<**>` is a `Rat` or `FatRat`, the resulting type will always be `Num` even if the base's type can hold the value without decaying.

For example,
```raku
say (2.FatRat ** 2.FatRat).WHAT;
# OUTPUT: (Num)

say (2.0 ** 2.0).WHAT;
# OUTPUT: (Num)

say (2.FatRat ** 2).WHAT;
# OUTPUT: (FatRat)

say (2.0 ** 2).WHAT;
# OUTPUT: (Rat)
```

This is highly counter-intuitive and seems to break the typical rules for when `Rat`s should decay. It also obviously produces unexpected results on `FatRat`s, which shouldn't decay at all.

**Comment by landyacht** [comment]
The simplest solution I can think of is to let the more "powerful" type decide the resulting type. So `(Int) ** (FatRat)` would result in a `FatRat`.

However, would it be desirable for `(Rat) ** (FatRat)` to produce a `FatRat`, or should the type of the base be more important than the type of the exponent in this case?

**Comment by jnthn** [comment]
There's currently no candidates involving rationals and exponentiation except for `multi sub infix:<**>(Rational:D \a, Int:D \b)`; for anything else, it's falling back to the `Real` default, which coerces the values to `Num` and does the operation on those. So it's not a matter of decay, but rather one of not even trying to do it with rationals.

Related to this, however, is `sqrt` (or any root operation) is not defined on `Rat` either. I wonder if the big integer library that underlies all of this supports it at all (finding any root, that is, because so far as I understand these things, that's the generality we'd need for fractional exponentiation).

**Comment by taboege** [comment]
> So `(Int) ** (FatRat)` would result in a `FatRat`.

This is not going to fly. Just consider `2 ** ½.FatRat` aka √2. This is not a rational number and no amount of fat can contain it in an integer numerator and denominator. With the current Raku types, you could only avoid decaying to a Num when the exponent is an integer (then it's just shorthand for multiplying or dividing by the mantissa) or when the mantissa happens to be a proper power that can absorb the rational exponent's denominator. Num is narrowest result type for exponentiation of rationals.

**Comment by AlexDaniel** [comment]
Yeah, it's a bit complicated. There are edge cases like fractional exponents (e.g. less than one) and negative ones. For example, right now with just Ints: 50**9 is an Int, 50**-9 is obviously a Rat, 50**-99 is a Num and 50**-999 is an Underflow exception. Arguably that makes sense, but I'm not sure. Then, 50**9999999999 is an overflow exception.

Maybe we need some sort of table for `**` with every edge case listed? Then we'd be able to see how proposed changes for `** Rat` will fit in.

**Comment by jnthn** [comment]
> This is not a rational number

True, and...

> There are edge cases 

I guess if we make it "only decay when it's irrational", that's a heck of an edge case... :)

**Comment by AlexDaniel** [comment]
> This is not a rational number and no amount of fat can contain it in an integer numerator and denominator

Yes, but `4**½` is just 2 and you don't even need a Rat to represent that. If ``**`` with Ints is allowed to degrade to Nums in some cases then maybe ``**`` with Rats can also do some kind of “best effor or Num”. Not arguing that it should be that way, but maybe maybe?

**Comment by MasterDuke17** [comment]
> Related to this, however, is `sqrt` (or any root operation) is not defined on `Rat` either. I wonder if the big integer library that underlies all of this supports it at all (finding any root, that is, because so far as I understand these things, that's the generality we'd need for fractional exponentiation).

https://github.com/libtom/libtommath/blob/develop/tommath.h#L422-L426
